sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/AssignmentUtils.scala

```diff
@@ -0,0 +1,275 @@
+            case colType: StructType =>
+              // build field expressions
+              val fieldExprs = colType.fields.zipWithIndex.map { case (field, ordinal) =>
+                GetStructField(col, ordinal, Some(field.name))
+              }
+
+              // recursively apply this method to nested fields
+              val updatedFieldExprs = applyUpdates(
+                matchedUpdates.map(update => update.copy(ref = update.ref.tail)),
+                colType.toAttributes,
+                fieldExprs,
+                addError,
+                newColPath)
+
+              // construct a new struct with updated field expressions
+              toNamedStruct(colType, updatedFieldExprs)
+
+            case otherType =>
+              addError(
+                "Updating nested fields is only supported for StructType but " +
+                s"${newColPath.quoted} is of type $otherType")
+              col
+          }
+
+        // if there are conflicting updates, throw an exception
+        // there are two illegal scenarios:
+        // - multiple updates to the same column
+        // - updates to a top-level struct and its nested fields (like a.b and a.b.c)
+        case matchedUpdates if hasExactMatch(matchedUpdates, col) =>
+          val conflictingColNames = matchedUpdates.map(update => (colPath ++ update.ref).quoted)
+          addError("Update conflicts for columns: " + conflictingColNames.distinct.mkString(", "))
+          col
+      }
+    }
+  }
+
+  private def toNamedStruct(structType: StructType, fieldExprs: Seq[Expression]): Expression = {
+    val namedStructExprs = structType.fields.zip(fieldExprs).flatMap { case (field, expr) =>
+      Seq(Literal(field.name), expr)
+    }
+    CreateNamedStruct(namedStructExprs)
+  }
+
+  private def hasExactMatch(updates: Seq[ColumnUpdate], col: NamedExpression): Boolean = {
+    updates.exists(isExactMatch(_, col))
+  }
+
+  private def isExactMatch(update: ColumnUpdate, col: NamedExpression): Boolean = {
+    update.ref match {
+      case Seq(namePart) if conf.resolver(namePart, col.name) => true
+      case _ => false
+    }
+  }
+
+  private def applyUpdate(
+      value: Expression,
+      col: Attribute,
+      addError: String => Unit,
+      colPath: Seq[String]): Expression = {
+
+    (value.dataType, col.dataType) match {
+      // no need to reorder inner fields or cast if types are equal ignoring nullability
+      case (valueType, colType) if valueType.sameType(colType) =>
+        TableOutputResolver.checkNullability(value, col, conf, addError, colPath)
+        validateAssignment(valueType, colType, addError, colPath)
+        value
+
+      case (valueType: StructType, colType: StructType) =>
+        TableOutputResolver.checkNullability(value, col, conf, addError, colPath)
+        val resolvedValue = TableOutputResolver.resolveStructType(
+          value, valueType, colType,
+          conf, addError, colPath)
+        resolvedValue.getOrElse(col)
```

Comment:
If cannot resolve it, why it is returning `col` instead of `value`?

Reply:
It doesn't really matter but makes more sense, let me change.

Reply:
Fixed. Let me resolve this to simplify reviews.